<img align="center" src="https://raw.githubusercontent.com/FUlyankin/Parsers/master/images%20/TOR.jpg" height="900" width="900"> 

# <center> Мне может и нечего скрывать, но чем я занимаюсь — это не ваше дело. Или "Что делать, если вас банят по ip. </center>

---------


Вопреки пропагандируемому мнению, Tor используется не только преступниками, педофилами и прочими нехорошими террористами. Это, мягко говоря, далеко не так. Активисты разного рода, журналисты, просто люди, любящие приватность составляют портрет пользователей Tor. Лично мне очень импонирует позиция разработчиков Tor, отвечающих на вопрос «а что, у вас есть, что скрывать?» фразой: «Нет, это не секрет – просто это не ваше дело».

У преступников есть и другие средства от смены личности и краденых устройств до специальных ботов и вирусов. Используя и пропагандируя Tor вы помогаете преступникам не более, чем используя интернет. Аналогичные вещи можно сказать и по поводу Telegram. Многие могут поспорить с такой позицией, но мы не для этого тут собрались...


## Как работает ТОР

Картинки и большая часть информации взяты со [статьи на geektimes.](https://geektimes.ru/post/277578/) Рекоменду почитать про ТОР [в интернатах :)](https://ru.wikipedia.org/wiki/Tor) Особенно про попытки сломать ТОР и поиск уязвимостей. Это поможет сформировать какое-то представления о нашей с вами анонимности в интернетах.

<img align="center" src="https://hsto.org/getpro/geektimes/post_images/f6c/2e7/99f/f6c2e799f6d101d30a07e355949cd15a.png" height="800" width="800"> 


<img align="center" src="https://hsto.org/getpro/geektimes/post_images/c8c/458/1af/c8c4581af03413ae7961b15f68a7d328.png" height="800" width="800"> 


**Входной, или сторожевой узел (нода)** – точка входа в сеть. Входные узлы выбираются из тех, что работают продолжительное время, и показали себя как стабильные и высокоскоростные.
**Промежуточный узел (нода)** – передаёт трафик от охранных к выходным. В результате первые не знают ничего о последних.
**Выходной узел (нода)** – точка выхода из сети, отправляет трафик к пункту назначения, который нужен клиенту.

Обычно безопасный метод запуска сторожевого или промежуточного узла — виртуальный сервер (DigitalOcean, EC2) – в этом случае операторы сервера увидят только зашифрованный трафик.

На операторах выходных узлов лежит особая ответственность. Поскольку они отправляют трафик в пункт назначения, все незаконные действия, совершаемые через Tor, будут связываться с выходным узлом. А это может привести к полицейским рейдам, уведомлениям о нелегальной деятельности и другим вещам. Встретите оператора выходного узла – скажите ему спасибо. Он того заслуживает.

<img align="center" src="https://raw.githubusercontent.com/FUlyankin/Parsers/master/images%20/tor_noda.jpg" height="650" width="650"> 


Вернёмся к нашим баранам. Мы – мирные собиратели данных, а не терористы. Иногда серверу надоедает общаться с одним и тем же человеком, делающим кучу запросов и сервер банит его. К сожалению, не только люди бывают невыносимыми... Придётся маскироваться... Мы будем использовать возможности Tor для обхода блокировок. 

**Ещё раз, ещё раз, мы не терористы! Правда...**

## 2. Добиваемся бана.

Дальнейший скрипт построен на материалах, созданных Димой Сергеевым.  Если хотите посмотреть первоисточник, а заодно и прочитать исследование о мемах, добро пожаловать в [Димин репозиторий](https://github.com/DmitrySerg/memology) с кучей кода и данных. Там много чего интересного. В том числе несколько парсеров... Также всем рекомендую посмотреть [выступление Димы на датафесте](https://it.mail.ru/video/933/), а также [его презенташку.](https://github.com/FUlyankin/Parsers/raw/master/Po_ai_774_pi_vychislyu_33.pdf) 

По аналогии с Димой добьёмся бана от http://knowyourmeme.com. Не будем особо заморачиваться с кодом парсера. Будем всё время обращаться к одной и той же странице до тех пор, пока нас не забанят.

Как только мы получим бан, сервер либо начнёт нам выдавать ошибку, либо сообщения о том, что мы были забанены. 

In [1]:
import requests
page = 'http://knowyourmeme.com/memes/researching/page/8'

In [2]:
k = 0
while k<5:
    response = requests.get(page)
    print(response)
    k += 1

<Response [403]>
<Response [403]>
<Response [403]>
<Response [403]>
<Response [403]>


In [3]:
response.content

b'<html>\r\n<head><title>403 Forbidden</title></head>\r\n<body bgcolor="white">\r\n<center><h1>403 Forbidden</h1></center>\r\n<hr><center>nginx/1.10.0 (Ubuntu)</center>\r\n</body>\r\n</html>\r\n'

Ошибка 403, доступ запрещён. Попробуем добиться доступа с помощью генерации случайных юзер-агентов. Предварительно не забываем установить библиотеку! 

In [4]:
from user_agent import generate_user_agent 

Сгенерим парочку юзер агентов:

In [5]:
generate_user_agent()

'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:48.0) Gecko/20100101 Firefox/48.0'

In [6]:
generate_user_agent()

'Mozilla/5.0 (Windows NT 10.0; WOW64; rv:48.0) Gecko/20100101 Firefox/48.0'

In [7]:
generate_user_agent()

'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:49.0) Gecko/20100101 Firefox/49.0'

Много разных юзер-агентов :) Пробуем обойти блокировку.

In [8]:
k = 0
while k<5:
    response = requests.get(page,headers = {'User-Agent':generate_user_agent()})
    print(response)
    k += 1

<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>


Благословенный 200 - ответ... Сервер снова нам отвечает. Убедимся, что респонс полон контента.

In [9]:
response.content

b'<h1>\nYour ip address\n<span style=\'color: red;\'>\n85.30.254.40\n</span>\nhas been banned.\n</h1>\n<p>\nIf you think this is a mistake then please contact\n<a href="mailto:bans@knowyourmeme.com?subject=85.30.254.40 banned">bans@knowyourmeme.com</a>\nimmediately and\n<strong>\n<em>\nmake sure to include your ip address.\n</em>\n</strong>\n</p>\n<h3 style=\'color: red; font-weight: bold;\'>\nThis should be obvious but if you don\'t include your ip address then there\'s no way to unban you.\n</h3>\n<h3>\n<em>\nDon\'t include a screenshot of this page or any other page in the email; it\'s just a waste of time.\n</em>\n</h3>\n<br>\n<hr>\n<br>\n<p>\n<strong>\nNOTE: If you\'re thinking of claiming that your ip address is actually 192.168.X.X and not the one listed above:\n</strong>\n</p>\n<p>\n<em>\n192.168.X.X is a private ip address for your local network and it\'s irrelevant to the outside world.\n</em>\n</p>\n'

Снова добиваемся бана.

In [10]:
k = 0
while k < 500:
    response = requests.get(page,headers = {'User-Agent':generate_user_agent()})
    print(response)
    k += 1

<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>
<Response [200]>


А что у нас лежит в респонсе?

In [11]:
response.content

b'<h1>\nYour ip address\n<span style=\'color: red;\'>\n85.30.254.40\n</span>\nhas been banned.\n</h1>\n<p>\nIf you think this is a mistake then please contact\n<a href="mailto:bans@knowyourmeme.com?subject=85.30.254.40 banned">bans@knowyourmeme.com</a>\nimmediately and\n<strong>\n<em>\nmake sure to include your ip address.\n</em>\n</strong>\n</p>\n<h3 style=\'color: red; font-weight: bold;\'>\nThis should be obvious but if you don\'t include your ip address then there\'s no way to unban you.\n</h3>\n<h3>\n<em>\nDon\'t include a screenshot of this page or any other page in the email; it\'s just a waste of time.\n</em>\n</h3>\n<br>\n<hr>\n<br>\n<p>\n<strong>\nNOTE: If you\'re thinking of claiming that your ip address is actually 192.168.X.X and not the one listed above:\n</strong>\n</p>\n<p>\n<em>\n192.168.X.X is a private ip address for your local network and it\'s irrelevant to the outside world.\n</em>\n</p>\n'

**Запускаем ТОР-браузер! Да, да! Прямо берём и открываем окошечко с ним.** Запускаем код ниже.

In [12]:
# Анонимизация 
# В консоли: pip install PySocks
import socks
import socket
socks.set_default_proxy(socks.SOCKS5, "localhost", 9150)
socket.socket = socks.socksocket
# ТОР по умолчанию идет через сокет 9150

Посмотрим в какой мы стране.

In [13]:
from urllib.request import urlopen
print(urlopen('http://icanhazip.com').read())

b'87.118.92.43\n'


[Выясним откуда мы сидим в интернете!](http://ru.smart-ip.net/geoip) Ого... Германия.

<img align="center" src="https://raw.githubusercontent.com/FUlyankin/Parsers/master/images%20/ip_adress.png" height="650" width="650"> 



In [14]:
response = requests.get(page,headers = {'User-Agent':generate_user_agent()})
response

<Response [200]>

In [14]:
response.content

b'<h1>\nYour ip address\n<span style=\'color: red;\'>\n85.248.227.164\n</span>\nhas been banned.\n</h1>\n<p>\nIf you think this is a mistake then please contact\n<a href="mailto:bans@knowyourmeme.com?subject=85.248.227.164 banned">bans@knowyourmeme.com</a>\nimmediately and\n<strong>\n<em>\nmake sure to include your ip address.\n</em>\n</strong>\n</p>\n<h3 style=\'color: red; font-weight: bold;\'>\nThis should be obvious but if you don\'t include your ip address then there\'s no way to unban you.\n</h3>\n<h3>\n<em>\nDon\'t include a screenshot of this page or any other page in the email; it\'s just a waste of time.\n</em>\n</h3>\n<br>\n<hr>\n<br>\n<p>\n<strong>\nNOTE: If you\'re thinking of claiming that your ip address is actually 192.168.X.X and not the one listed above:\n</strong>\n</p>\n<p>\n<em>\n192.168.X.X is a private ip address for your local network and it\'s irrelevant to the outside world.\n</em>\n</p>\n'

Бан успешно снят! Респонс снова полон контента. Wow! 

## Мораль

Попробуйте накачивать свой парсер примочками постепенно. 

1. Нужно быть терпеливым, time.sleep(n). Выясните ограничения. Обращайтесь к серверу пореже и, возможно, он вас не забанит.

2. Попытайтесь быть похожим на человка. Добавьте дополнительных полей в свои запросы. 

3. Если это всё не помогает, используйте ТОР. 

Каждая новая примочка замедляет работу кода. 

Так уж вышло, что пара по обходу блокировок пришлась на тот момент, когда правительство Украины заблокировало у себя в стране довольно большое количество Российских сервисов от вконтакте до яндекса. Топ Украинского Appstore выглядит сейчас вот так: 


<img align="center" src="https://raw.githubusercontent.com/FUlyankin/Parsers/master/images%20/ukraina.jpg
" height="300" width="300">

-----------------

Ещё всем рекомендую посмотреть сериал ["Чёрное Зеркало"](https://www.kinopoisk.ru/film/655800/) о том как технологии до неузнаваемости меняют наш мир. Каждая серия в нём это новая история. Они никак не связаны. Некоторые производят очень сильное впечатление... 
